File: src/main/java/com/now/nowbot/service/MessageServiceImpl/NominationService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.model.JsonData.BeatMap;
import com.now.nowbot.model.JsonData.BeatMapSet;
import com.now.nowbot.model.JsonData.Discussion;
import com.now.nowbot.model.JsonData.DiscussionDetails;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.ImageService;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.OsuApiService.OsuBeatmapApiService;
import com.now.nowbot.service.OsuApiService.OsuDiscussionApiService;
import com.now.nowbot.service.OsuApiService.OsuUserApiService;
import com.now.nowbot.throwable.ServiceException.NominationException;
import com.now.nowbot.util.Instructions;
import jakarta.annotation.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.util.*;
import java.util.regex.Matcher;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service("NOMINATION")
public class NominationService implements MessageService<Matcher> {
    private static final Logger log = LoggerFactory.getLogger(NominationService.class);

    @Resource
    OsuBeatmapApiService osuBeatmapApiService;
    @Resource
    OsuUserApiService osuUserApiService;
    @Resource
    OsuDiscussionApiService osuDiscussionApiService;
    @Resource
    ImageService imageService;

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<Matcher> data) throws Throwable {

        var matcher = Instructions.NOMINATION.matcher(messageText);
        if (! matcher.find()) return false;

        data.setValue(matcher);
        return true;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        var from = event.getSubject();
        long sid;
        boolean isSID = true;

        String mode = matcher.group("mode");

        if (Objects.nonNull(mode) && (Objects.equals(mode, "b") || Objects.equals(mode, "bid"))) {
            isSID = false;
        }

        try {
            sid = Long.parseLong(matcher.group("sid"));
        } catch (NumberFormatException e) {
            throw new NominationException(NominationException.Type.N_Instructions);
        }

        var data = parseData(sid, isSID, osuBeatmapApiService, osuDiscussionApiService, osuUserApiService);

        byte[] image;

        try {
            image = imageService.getPanelN(data);
        } catch (Exception e) {
            log.error("提名信息：渲染失败", e);
            throw new NominationException(NominationException.Type.N_Render_Failed);
        }

        try {
            from.sendImage(image);
        } catch (Exception e) {
            log.error("提名信息：发送失败", e);
            throw new NominationException(NominationException.Type.N_Send_Error);
        }
    }

    public static Map<String, Object> parseData(long sid, boolean isSID, OsuBeatmapApiService beatmapApiService, OsuDiscussionApiService discussionApiService, OsuUserApiService userApiService
    ) throws NominationException {
        BeatMapSet s;
        Discussion d;
        final List<DiscussionDetails> details;
        final List<DiscussionDetails> discussions;
        final List<DiscussionDetails> hypes;
        Map<String, Object> more = new HashMap<>();

        if (isSID) {
            try {
                s = beatmapApiService.getBeatMapSetInfo(sid);
            } catch (WebClientResponseException.NotFound | HttpClientErrorException.NotFound e) {
                try {
                    var b = beatmapApiService.getBeatMapInfoFromDataBase(sid);
                    sid = b.getSetID();
                    s = beatmapApiService.getBeatMapSetInfo(sid);
                } catch (WebClientResponseException.NotFound | HttpClientErrorException.NotFound e1) {
                    throw new NominationException(NominationException.Type.N_Map_NotFound);
                } catch (Exception e1) {
                    log.error("提名信息：谱面获取失败", e1);
                    throw new NominationException(NominationException.Type.N_Map_FetchFailed);
                }
            } catch (WebClientResponseException.BadGateway | WebClientResponseException.ServiceUnavailable e) {
                throw new NominationException(NominationException.Type.N_API_Unavailable);
            } catch (Exception e) {
                log.error("提名信息：谱面获取失败", e);
                throw new NominationException(NominationException.Type.N_Map_FetchFailed);
            }
        } else {
            try {
                var b = beatmapApiService.getBeatMapInfoFromDataBase(sid);
                sid = b.getSetID();
                s = beatmapApiService.getBeatMapSetInfo(sid);
            } catch (WebClientResponseException.NotFound | HttpClientErrorException.NotFound e) {
                throw new NominationException(NominationException.Type.N_Map_NotFound);
            } catch (WebClientResponseException.BadGateway | WebClientResponseException.ServiceUnavailable e) {
                throw new NominationException(NominationException.Type.N_API_Unavailable);
            } catch (Exception e) {
                log.error("提名信息：谱面获取失败", e);
                throw new NominationException(NominationException.Type.N_Map_FetchFailed);
            }
        }

        if (Objects.nonNull(s.getCreatorData())) {
            s.getCreatorData().parseFull(userApiService);
        }



        try {
            d = discussionApiService.getBeatMapSetDiscussion(sid);
        } catch (Exception e) {
            log.error("提名信息：讨论区获取失败", e);
            throw new NominationException(NominationException.Type.N_Discussion_FetchFailed);
        }

        //插入难度名
        if (Objects.nonNull(s.getBeatMaps())) {
            Map<Long, String> diffs = s.getBeatMaps().stream().collect(
                    Collectors.toMap(BeatMap::getBeatMapID, BeatMap::getDifficultyName)
            );

            d.addDifficulty4DiscussionDetails(diffs);
        }

        //获取 hypes 和 discussions 列表
        {
            //这两个list需要合并起来
            details = Stream.of(d.getDiscussions(), d.getIncludedDiscussions())
                    .filter(Objects::nonNull)
                    .flatMap(Collection::stream)
                    .distinct()
                    .toList();

            hypes = details.stream().filter(i -> {
                var t = i.getMessageType();
                return t.equals(DiscussionDetails.MessageType.hype) || t.equals(DiscussionDetails.MessageType.praise);
            }).toList();

            var dis = details.stream().filter(i -> {
                var t = i.getMessageType();
                return t.equals(DiscussionDetails.MessageType.problem) || t.equals(DiscussionDetails.MessageType.suggestion);
            }).toList();

            discussions = Discussion.toppingUnsolvedDiscussionDetails(dis);
        }
        //这一部分提供额外信息
        {
            int hostCount = 0;
            int guestCount = 0;
            int problemCount = 0;
            int suggestCount = 0;
            int notSolvedCount = 0;
            int hypeCount = 0;
            int praiseCount = 0;
            String maxSR = "";
            String minSR = "";
            int totalLength = 0;
            List<Float> SRList = new ArrayList<>();

            for (var i : details) {
                switch (i.getMessageType()) {
                    case problem -> problemCount ++;
                    case suggestion -> suggestCount ++;
                    case hype -> hypeCount ++;
                    case praise -> praiseCount ++;
                }

                if (i.getCanBeResolved() && !i.getResolved()) {
                    notSolvedCount ++;
                }
            }

            var bs = s.getBeatMaps();

            //初始化星数
            float maxStarRating = 0;
            float minStarRating = 0;

            if (Objects.nonNull(bs) && !bs.isEmpty()) {
                var f = bs.getFirst();
                totalLength = f.getTotalLength();
                maxStarRating = f.getStarRating();
                minStarRating = maxStarRating;
            }

            if (Objects.nonNull(bs)) {
                for (var b : bs) {
                    if (Objects.equals(s.getCreatorID(), b.getMapperID())) {
                        hostCount ++;
                    } else {
                        guestCount ++;
                    }

                    if (b.getStarRating() > maxStarRating) maxStarRating = b.getStarRating();
                    if (b.getStarRating() < minStarRating) minStarRating = b.getStarRating();

                    SRList.add(b.getStarRating());
                }

                var maxStarRatingInt = (int) Math.floor(maxStarRating);
                var minStarRatingInt = (int) Math.floor(minStarRating);

                maxSR = String.valueOf(maxStarRatingInt);
                minSR = String.valueOf(minStarRatingInt);

                if (maxStarRating - maxStarRatingInt >= 0.5) maxSR += '+';
                //if (minStarRating - minStarRatingInt >= 0.5) minSR += '+';

                //单难度
                if (bs.size() <= 1) minSR = "";
            }

            //其他
            String[] tags = s.getTags().split(" ");

            more.put("hostCount", hostCount);
            more.put("guestCount", guestCount);
            more.put("totalCount", hostCount + guestCount);
            more.put("maxSR", maxSR);
            more.put("minSR", minSR);
            more.put("SRList", SRList.stream().sorted(
                    Comparator.comparingDouble(Float::doubleValue).reversed()
            ).toList());
            more.put("totalLength", totalLength);
            more.put("tags", tags);
            more.put("problemCount", problemCount);
            more.put("suggestCount", suggestCount);
            more.put("notSolvedCount", notSolvedCount);
            more.put("hypeCount", hypeCount);
            more.put("praiseCount", praiseCount);
        }

        var n = new HashMap<String, Object>();
        n.put("beatmapset", s);
        n.put("discussion", discussions);
        n.put("hype", hypes);
        n.put("more", more);
        n.put("users", d.getUsers());

        return n;
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/NominationService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.dao.ServiceCallStatisticsDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.model.osu.Beatmapset
import com.now.nowbot.model.osu.Discussion
import com.now.nowbot.model.osu.Discussion.Companion.toppingUnsolvedDiscussionDetails
import com.now.nowbot.model.osu.DiscussionDetails
import com.now.nowbot.model.osu.OsuUser
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.qq.message.MessageChain
import com.now.nowbot.qq.message.MessageChain.MessageChainBuilder
import com.now.nowbot.qq.tencent.TencentMessageService
import com.now.nowbot.service.ImageService
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.messageServiceImpl.NominationService.NominationParam
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService
import com.now.nowbot.service.osuApiService.OsuDiscussionApiService
import com.now.nowbot.service.osuApiService.OsuUserApiService

import com.now.nowbot.throwable.botRuntimeException.IllegalStateException
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.OfficialInstruction
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.regex.Matcher
import kotlin.math.floor

@Service("NOMINATION") class NominationService(
    private val beatmapApiService: OsuBeatmapApiService,
    private val userApiService: OsuUserApiService,
    private val discussionApiService: OsuDiscussionApiService,
    private val imageService: ImageService,
    private val dao: ServiceCallStatisticsDao,
) : MessageService<NominationParam>, TencentMessageService<NominationParam> {

    data class NominationParam(
        val beatmapset: Beatmapset,
        val discussions: List<DiscussionDetails>,
        val hype: List<DiscussionDetails>,
        val more: Map<String, Any>,
        val users: List<OsuUser>
    ) {
        fun toMap(): Map<String, Any> {
            return mapOf(
                "beatmapset" to beatmapset,
                "discussion" to discussions,
                "hype" to hype,
                "more" to more,
                "users" to users,
            )
        }
    }

    override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<NominationParam>,
    ): Boolean {
        val matcher = Instruction.NOMINATION.matcher(messageText)
        if (!matcher.find()) return false

        val param = getNominationParam(event, matcher)

        data.value = param
        return true
    }

    override fun handleMessage(event: MessageEvent, param: NominationParam): ServiceCallStatistic? {
        val image: ByteArray = imageService.getPanel(param.toMap(), "N")

        try {
            event.replyAsync(image)
        } catch (e: Exception) {
            log.error("提名信息：发送失败", e)
            throw IllegalStateException.Send("提名信息")
        }

        return ServiceCallStatistic.builds(event,
            beatmapIDs = param.beatmapset.beatmaps?.map { it.beatmapID }.orEmpty(),
            beatmapsetIDs = listOf(param.beatmapset.beatmapsetID),
            userIDs = param.users.map { it.userID },
            modes = param.beatmapset.beatmaps?.map { it.mode }?.distinct()
        )
    }

    override fun accept(event: MessageEvent, messageText: String): NominationParam? {
        val matcher = OfficialInstruction.NOMINATION.matcher(messageText)
        if (!matcher.find()) return null

        val param = getNominationParam(event, matcher)

        return param
    }

    override fun reply(event: MessageEvent, param: NominationParam): MessageChain? {
        val image: ByteArray =  imageService.getPanel(param.toMap(), "N")

        return MessageChainBuilder().addImage(image).build()
    }

    private fun getNominationParam(event: MessageEvent,
        matcher: Matcher,
    ): NominationParam {
        val (beatmapset, _) = beatmapApiService.getBeatmapsetAndTopBeatmapFromAnyID(matcher) { dao.getLastBeatmapID(event) }

        val discussion = runCatching {
            discussionApiService.getBeatmapsetDiscussion(beatmapset.beatmapsetID)
        }.onFailure { e ->
            log.error("提名信息：讨论区获取失败", e)
            throw IllegalStateException.Fetch("讨论区")
        }.getOrThrow()

        beatmapset.creatorData?.let {
            try {
                beatmapset.creatorData = userApiService.getOsuUser(it.userID, it.currentOsuMode)
            } catch (_: Exception) {

            }
        }

        beatmapset.beatmaps!!.forEach { beatmapApiService.extendBeatmapTag(it) }

        val diffs = beatmapset.beatmaps!!.associate { it.beatmapID to it.difficultyName }
        discussion.addDifficulty4DiscussionDetails(diffs)

        return getParam(
            beatmapset,
            discussion,
        )
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(NominationService::class.java)

        fun getParam(
            beatmapset: Beatmapset,
            discussion: Discussion,
        ): NominationParam {
            val details: List<DiscussionDetails>
            val discussions: List<DiscussionDetails>
            val hypes: List<DiscussionDetails>
            val more: Map<String, Any>

            // 获取 hypes 和 discussions 列表
            // 这两个list需要合并起来
            run {
                details = (discussion.discussions + discussion.includedDiscussions).distinct()
                hypes = details.filter { i: DiscussionDetails ->
                    i.messageType.isComment()
                }

                val dis = details.filter { i: DiscussionDetails ->
                    i.messageType.isCheck()
                }

                discussions = dis.toppingUnsolvedDiscussionDetails()
            }

            // 这一部分提供额外信息
            run {
                var problemCount = 0
                var suggestCount = 0
                var notSolvedCount = 0
                var hypeCount = 0
                var praiseCount = 0

                for (i in details) {
                    when (i.messageType) {
                        DiscussionDetails.DiscussionType.PROBLEM -> problemCount++
                        DiscussionDetails.DiscussionType.SUGGESTION -> suggestCount++
                        DiscussionDetails.DiscussionType.HYPE -> hypeCount++
                        DiscussionDetails.DiscussionType.PRAISE -> praiseCount++
                        else -> {}
                    }

                    if (i.canBeResolved && !i.resolved) {
                        notSolvedCount++
                    }
                }

                val bs = beatmapset.beatmaps.orEmpty()

                val stars = bs.map { it.starRating }

                val maxStar = stars.max()
                val minStar = stars.min()
                val totalLength = bs.sumOf { it.totalLength }

                val hostCount = bs.count { it.mapperIDs.contains(beatmapset.creatorID) }
                val guestCount = bs.size - hostCount

                val maxSR = floor(maxStar).toString() + if (maxStar - floor(maxStar) >= 0.5) "+" else ""
                val minSR = if (bs.size <= 1) "" else floor(minStar).toString()

                // 其他
                val tags = beatmapset.tags.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                more = mapOf(
                    "host_count" to hostCount,
                    "guest_count" to guestCount,
                    "total_count" to (hostCount + guestCount),
                    "max_star" to maxSR,
                    "min_star" to minSR,
                    "stars" to stars.sortedDescending(),
                    "total_length" to totalLength,
                    "tags" to tags,
                    "problem_count" to problemCount,
                    "suggest_count" to suggestCount,
                    "not_solved_count" to notSolvedCount,
                    "hype_count" to hypeCount,
                    "praise_count" to praiseCount,
                )
            }

            return NominationParam(
                beatmapset, discussions, hypes, more, discussion.users
            )
        }
    }
}
